javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -11,39 +11,41 @@
 
 import java.io.Serializable;
 import java.util.*;
-import java.util.Set;
 import java.util.function.*;
 import java.util.stream.Collector;
 
+import static javaslang.collection.BitMappedTrie.lastDigit;
 import static javaslang.collection.Collections.*;
 
 /**
- * Vector is the default Seq implementation. It provides the best performance in between Array (with constant time element access)
- * and List (with constant time element addition).
+ * Vector is the default Seq implementation that provides effectively constant time access to any element.
+ * Many other operations (e.g. `tail`, `drop`, `slice`) are also effectively constant.
  *
  * @param <T> Component type of the Vector.
- * @author Ruslan Sennov
+ * @author Ruslan Sennov, Pap Lőrinc
+ * @see BitMappedTrie
  * @since 2.0.0
  */
 public final class Vector<T> implements Kind1<Vector<?>, T>, IndexedSeq<T>, Serializable {
     private static final long serialVersionUID = 1L;
 
-    private static final Vector<?> EMPTY = new Vector<>(HashArrayMappedTrie.empty());
+    private static final Vector<?> EMPTY = new Vector<>(BitMappedTrie.empty());
 
-    private final HashArrayMappedTrie<Integer, T> trie;
-    private final int indexShift;
-
-    private static <T> Vector<T> wrap(HashArrayMappedTrie<Integer, T> trie) {
-        return trie.isEmpty() ? empty() : new Vector<>(trie);
+    final BitMappedTrie<T> trie;
+    private Vector(BitMappedTrie<T> trie) {
+        assert (EMPTY == null) || (trie.length() > 0);
+        this.trie = trie;
     }
 
-    private Vector(HashArrayMappedTrie<Integer, T> trie) {
-        this(0, trie);
+    @SuppressWarnings("ObjectEquality")
+    private Vector<T> wrap(BitMappedTrie<T> trie) {
+        return (trie == this.trie) ? this
```

Comment:
Currently we have only one case where `trie == this.trie` and `trie.length() > 0`, namely `trie.drop(n)` where `n <= 0`:

```java
wrap(trie.append(element));  // trie.append(element) != trie
wrap(trie.drop(n));          // trie.drop(n) == trie <=> n <= 0
wrap(trie.prepend(element)); // trie.prepend(element) != trie
wrap(trie.take(n));          // trie.take(n) == trie <=> trie is empty
```

I thought of merging `wrap(BMT)` and `ofAll(BMT)` to one static method `wrap(BMT self, BMT that)` but that is not possible for the other static `ofAll` methods (without passing `null` as self).

I would like to rename `ofAll(BMT)` to `wrap(BMT)`. In order to resolve the name clash we make the current `wrap(BMT)` static and pass the `this` instance as `self` parameter: `wrap(self, that)`.